content/docs/concepts/resources/functions.md

```diff
@@ -131,34 +131,187 @@ variables:
+- `pluginDownloadURL`: Pass a URL from which the provider plugin should be fetched. This may be necessary for e.g. third-party packages (those not hosted at [https://get.pulumi.com](https://get.pulumi.com)).
 
-The `provider` option gives an explicit provider to use when running the invoked function. This is useful, for example, if you want to invoke a function in each of a set of AWS regions.
+- `provider`: Pass an [explicitly configured provider](/docs/concepts/resources/providers/#explicit-provider-configuration) to use for this function call, instead of using the default provider. This is useful, for example, if you want to invoke a function in each of a set of AWS regions.
 
-The `version` option specifies an exact version for the provider plugin. This can be used when you need to pin to a specific version to avoid a backward-incompatible change.
+- `version`: Pass a provider plugin version that should be used when invoking the function.
 
-The `pluginDownloadURL` option gives a URL for fetching the provider plugin. It may be necessary to supply this for third-party packages (those not hosted at [https://get.pulumi.com](https://get.pulumi.com)).
+- `async`: _This option is deprecated and will be removed in a future release_
+
+### Dependencies and ordering
+
+While the direct and output forms of a provider function are equivalent in terms
+of the results they produce when invoked, they differ in how they interact with
+the rest of the Pulumi program and the order in which they may be executed.
+Specifically:
+
+- Direct form invocations execute just like any other function call in the language. Since they do not accept Pulumi `Input`s nor return Pulumi `Output`s, they are not tracked by the Pulumi engine and do not participate in the dependency graph.
+
+- Output form invocations, on the other hand, are tracked by the Pulumi engine and participate in the dependency graph. This means, for example, that Pulumi will ensure that input resources are created or updated before an invocation and that the invocation is executed before dependent resources are created or updated.
+
+In general, you should prefer the output form of a provider function when you need to ensure that the function is executed in the correct order with respect to other resources in your program. This applies particularly if you wish to simulate something like the [`dependsOn` resource option](/docs/concepts/options/dependson/), where you want to ensure that a resource is created before the function is invoked, without necessarily passing the resource as an argument to the function. While there is no `dependsOn` [invoke option](#invoke-options) for provider functions, you can achieve the same effect using Pulumi's [`all`](/docs/concepts/inputs-outputs/all/) function and `Output`'s [`apply` method](/docs/concepts/inputs-outputs/apply/):
```

Comment:
This wording feels a bit squishy to me. As I understand it, it's not that you should _generally prefer_ the output form when you need to ensure that the function is executed in the correct order with respect to other resources -- it's that that's your only option. Right? (In other words, you _can't_ use the direct form to invoke a function in relation to the output of a resource because resource outputs don't produce plain values.) The next sentence is also a bit confusing in that it begins with "this applies" (i.e. presumably, "using the output form applies") when you want to ensure that a resource is created before the function is invoked _without passing the resource as an argument_ -- but since doing so is actually impossible, it can't apply either.

IMO it'd be clearer and more direct to say that when a function accepts one or more output arguments, Pulumi will resolve the outputs before invoking the function, and if you need to specify a dependency on outputs that the function doesn't depend on directly, you can use `pulumi.all` to tell Pulumi to resolve those outputs first, then invoke the function yourself, as in the following example. (Or similar.) 

Mentioning `dependsOn` probably makes sense too, given we're adding this in response to a request for that property. (Which I've actually wished for myself!) Do you know if we're planning to do this, though? If we are, we might instead want to link to that issue from this page and then position this approach as an option to be used in the meantime.

Reply:
Good shout -- I've gone for some stronger wording on this and would be grateful to get your thoughts! In terms of linking an issue -- for now, this is the resolution to that issue, precisely because the sharing of `InvokeOptions` across the two forms when only one (the output form) actually supports the notion, is problematic. Open to thoughts on this though.